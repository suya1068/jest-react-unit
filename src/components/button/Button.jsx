import classnames from "classnames";
import "./Button.scss";
import React, { PureComponent } from "react";
import PropTypes from "prop-types";

class Button extends PureComponent {
    render() {
        const { children, theme, ...rest } = this.props;
        return (
            <button className={classnames("u-button", Button.theme[theme])}>
                { children }
            </button>
        );
    }
}

Button.theme = {
    "primary": "u-button-primary",
    "secondary": "u-button-secondary",
    "positive": "u-button-positive",
    "negative": "u-button-negative"
};

Button.propTypes = {
    children: PropTypes.node.isRequired,
    theme: PropTypes.oneOf(Object.keys(Button.theme))
};

Button.defaultProps = {
    theme: null
};

export default Button;
