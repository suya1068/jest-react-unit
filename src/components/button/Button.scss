$u-button-primary: #2185d0;
$u-button-primary-hover: #1678c2;
$u-button-secondary: #1b1c1d;
$u-button-secondary-hover: #27292a;
$u-button-positive: #21ba45;
$u-button-positive-hover: #16ab39;
$u-button-negative: #db2828;
$u-button-negative-hover: #d01919;

.u-button {
    position: relative;
    display: inline-block;
    min-height: 1em;
    margin: 0 .25em 0 0;
    padding: 0.78571429em 1.5em 0.78571429em;
    background-color: #e0e1e2;
    border: none;
    border-radius: .28571429rem;
    box-shadow: 0 0 0 1px transparent inset, 0 0 0 0 rgba(34,36,38,.15) inset;
    outline: 0;
    line-height: 1em;
    vertical-align: baseline;
    color: rgba(0, 0, 0, 0.6);
    font-weight: 700;
    font-size: 1rem;
    text-align: center;
    text-decoration: none;
    cursor: pointer;
    user-select: none;
    transition: opacity 0.1s ease, background-color 0.1s ease, color 0.1s ease, box-shadow 0.1s ease;
    will-change: '';
    -webkit-tap-highlight-color: transparent;

    &:hover,
    &:focus {
        background-color: #cacbcd;
    }

    // theme
    &-primary {
        background-color: $u-button-primary;
        color: #fff;
        text-shadow: none;
        background-image: none;

        &:hover,
        &:focus {
            background-color: $u-button-primary-hover;
        }
    }

    &-secondary {
        background-color: $u-button-secondary;
        color: #fff;
        text-shadow: none;
        background-image: none;

        &:hover,
        &:focus {
            background-color: $u-button-secondary-hover;
        }
    }

    &-positive {
        background-color: $u-button-positive;
        color: #fff;
        text-shadow: none;
        background-image: none;

        &:hover,
        &:focus {
            background-color: $u-button-positive-hover;
        }
    }

    &-negative {
        background-color: $u-button-negative;
        color: #fff;
        text-shadow: none;
        background-image: none;

        &:hover,
        &:focus {
            background-color: $u-button-negative-hover;
        }
    }
}
