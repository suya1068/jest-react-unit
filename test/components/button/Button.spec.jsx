import React from "react";
import Button from "src/components/button/Button";

describe("Button Component", () => {
    it("충돌없이 랜더링한다.", () => {
        const tree = shallow(<Button>버튼</Button>);
        expect(tree).toMatchSnapshot();
    });

    it("테마를 설정한다.", () => {
        const tree = shallow(<Button theme="primary">버튼</Button>);
        expect(tree.find('.u-button').hasClass(Button.theme.primary)).toBe(true);

        tree.setProps({ theme: "secondary" });
        expect(tree.find('.u-button').hasClass(Button.theme.secondary)).toBe(true);

        tree.setProps({ theme: "positive" });
        expect(tree.find('.u-button').hasClass(Button.theme.positive)).toBe(true);

        tree.setProps({ theme: "negative" });
        expect(tree.find('.u-button').hasClass(Button.theme.negative)).toBe(true);
    });
});

/*
    // https://semantic-ui.com/elements/button.html
    1. 유효하지 않은 테마를 설정하면, 타입에러가 발생한다.
    2. 유효하지 않은 사이즈를 설정하면, 타입에러가 발생한다.
    3. 버튼을 클릭하면, 콜백함수가 실행된다.

    // 기본 속성
    * type - one of ['button', 'submit']. default 'button'
    * disabled - one of [true, false]. default false
    * autofocus - one of [true, false]. default false

    // 커스텀 속성
    * theme - one of ['primary', 'secondary', 'positive', 'negative']. default null
    * size - one of ['tiny', 'small', 'large', 'big']. default null (? { width: number, height: number })
    * shape - one of ['circle', 'round', 'rectangle']. default 'round'
    * transparent - one of [true, false]. default false
*/

/*
    function Button() {}

    new Button();

    new Button({
        theme: 'primary'
    });

    new Button({
        theme: 'primary',
        size: 'small'
    });

    new Button({
        theme: 'primary',
        size: 'small',
        transparent: true
    });

    new Button({
        theme: 'primary',
        size: 'small',
        disabled: true
    });
*/
