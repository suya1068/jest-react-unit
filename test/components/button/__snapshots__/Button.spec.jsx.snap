// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component 충돌없이 랜더링한다. 1`] = `
<button
  className="u-button"
>
  버튼
</button>
`;
